sources/web/datalab/polymer/modules/github-file-manager/github-file-manager.ts

```diff
@@ -0,0 +1,268 @@
+          return this._ghFileToContentString(response);
+        });
+  }
+
+  public async getRootFile() {
+    return this.get(new DatalabFileId('/', FileManagerType.GITHUB));
+  }
+
+  public saveText(_file: DatalabFile, _content: string): Promise<DatalabFile> {
+    throw new UnsupportedMethod('saveText', this);
+  }
+
+  public list(fileId: DatalabFileId): Promise<DatalabFile[]> {
+    const pathParts = fileId.path.split('/').filter((part) => !!part);
+    if (pathParts.length === 0) {
+      // No org/user specified. This would mean we should list all of them,
+      // but we know that's too many, so we return an empty list.
+      return Promise.resolve([]);
+    } else if (pathParts.length === 1) {
+      // Only the username or org was specified, list their repos
+      const githubPath = '/users/' + pathParts[0] + '/repos';
+      return this._githubApiPathRequest(githubPath)
+          .then((response: GhRepoResponse[]) => {
+            return this._ghReposResponseToDatalabFiles(response);
+          });
+    } else {
+      // If at least two path components were specified, then we have
+      // a username and a project. Everything after that, if specified,
+      // are folders or files under that.
+      const githubPath = '/repos/' + pathParts.slice(0,2).join('/') +
+        '/contents/' + pathParts.slice(2).join('/');
+      return this._githubApiPathRequest(githubPath)
+          .then((response: GhDirEntryResponse[]) => {
+            return this._ghDirEntriesResponseToDatalabFiles(response);
+          });
+    }
+  }
+
+  public create(_fileType: DatalabFileType, _containerId: DatalabFileId, _name: string):
+      Promise<DatalabFile> {
+    throw new UnsupportedMethod('create', this);
+  }
+
+  public rename(_oldFileId: DatalabFileId, _name: string, _newContainerId?: DatalabFileId):
+      Promise<DatalabFile> {
+    throw new UnsupportedMethod('rename', this);
+  }
+
+  public delete(_fileId: DatalabFileId): Promise<boolean> {
+    throw new UnsupportedMethod('delete', this);
+  }
+
+  public copy(_fileId: DatalabFileId, _destinationDirectoryId: DatalabFileId): Promise<DatalabFile> {
+    throw new UnsupportedMethod('copy', this);
+  }
+
+  public async getEditorUrl(fileId: DatalabFileId) {
+    return Utils.getHostRoot() + '/editor?file=' + fileId.toQueryString();
+  }
+
+  public async getNotebookUrl(fileId: DatalabFileId): Promise<string> {
+    return location.protocol + '//' + location.host +
+        '/notebook?file=' + fileId.toQueryString();
+  }
+
+  public pathToPathHistory(path: string): DatalabFile[] {
+    const pathParts = path.split('/').filter((part) => !!part);
+    const files: DatalabFile[] = [];
+    for (let p = 0; p < pathParts.length; p++) {
+      files[p] = this._ghPathPartsToDatalabFile(pathParts.slice(0, p + 1));
+    }
+    return files;
+  }
+
+  // We don't know if the tyep of the item is actually a directory without
+  // querying the github API, so we assume every component is a dir.
```

Comment:
This isn't always going to be correct though right? I'd add a TODO here at least.